apps/webapp/shared/select-token-modal.tsx

```diff
@@ -1,42 +1,29 @@
 'use client';
 
-import { useMemo, useState } from 'react';
-import { Dialog, DialogContent, DialogHeader, DialogPrimitive, DialogTrigger, Input } from 'ui';
-import { FilledImage } from '../filled-image';
 import { MagnifyingGlassIcon } from '@radix-ui/react-icons';
-import { assets } from '../../app/send/constants';
-import { Asset } from '../../types/asset';
-import { formatNumber } from '../../utils';
+import { useState } from 'react';
+import { Dialog, DialogContent, DialogHeader, DialogPrimitive, DialogTrigger, Input } from 'ui';
+import { FilledImage } from './filled-image';
+import { Asset, assets } from 'penumbra-constants';
+import { formatNumber } from '../utils';
+import { useSortedAssets } from '../hooks/sorted-asset';
 
 interface SelectTokenModalProps {
   asset: Asset;
   setAsset: (asset: Asset) => void;
 }
 
-export const SelectTokenModal = ({ asset, setAsset }: SelectTokenModalProps) => {
+export default function SelectTokenModal({ asset, setAsset }: SelectTokenModalProps) {
   const [search, setSearch] = useState('');
 
-  const filteredAsset = useMemo(() => {
-    const sortedAsset = [...assets].sort((a, b) => {
-      // Sort by balance in descending order (largest to smallest).
-      if (a.balance !== b.balance) return b.balance - a.balance;
-      // If balances are equal, sort by asset name in ascending order
-      return a.name.localeCompare(b.name);
-    });
-
-    // If no search query is provided, return the sorted assets directly.
-    if (!search) return sortedAsset;
-
-    // Filter the sorted assets based on a case-insensitive search query.
-    return sortedAsset.filter(asset => asset.name.toLowerCase().includes(search.toLowerCase()));
-  }, [search]);
+  const sortedAssets = useSortedAssets('balance', search);
 
   return (
     <Dialog>
-      <DialogTrigger asChild>
-        <div className='flex items-center gap-2 bg-light-brown px-6 py-[6px] rounded-lg cursor-pointer'>
-          <FilledImage src={asset.icon} alt='Asset' className='w-6 h-6' />
-          <p className='text-base font-bold text-light-grey'>{asset.name}</p>
+      <DialogTrigger disabled={!assets.length}>
+        <div className='flex items-center justify-center gap-2 bg-light-brown w-[100px] h-9 rounded-lg'>
+          {asset.icon && <FilledImage src={asset.icon} alt='Asset' className='w-6 h-6' />}
```

Comment:
I'm not sure if we can ever rely upon `asset.icon`. But I suppose it's best for now. Later, we'll want to add a default state if we don't have the icon.